### PR TITLE
[Large data feature importance] Add individual feature importance

### PR DESCRIPTION
**Pull request by vinuthakaranth**
## Description
This PR has below changes:

1. Adds feature importance for local explanations
2. Initially bubble chart will be displayed for large data
3. On clicking bubble - scatter plot will be displayed with points in the bubble
4. On clicking each point - local explanations for that point will be displayed
5. View as absolute values toggle
6. Class importance weights based on model type
7. Slider to change num. of features on highchart
8. Revert to bubble chart when in scatter plot of a specific bubble

Individual feature importance for classification:
FIclassExp

Individual feature importance for multi classification:
FIMultiClassExp

Individual feature importance for regression:
FIRegressionExp

Not large data experience will remain the same:
FIIOldExp

## Checklist

- [x] I have added screenshots above for all UI changes.
- [ ] I have added e2e tests for all UI changes.
- [ ] Documentation was updated if it was needed.

### CODE REVIEW

File: libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/FeatureImportances.tsx [hunk 1]
```diff
@@ -117,21 +116,35 @@ export class FeatureImportancesTab extends React.PureComponent<
...
+              selectedWeightVector={this.props.selectedWeightVector}
+              weightOptions={this.props.weightVectorOptions}
+              weightLabels={this.props.weightVectorLabels}
+              onWeightChange={this.props.onWeightVectorChange}
+              selectedCohort={this.context.selectedErrorCohort}
+              modelType={this.props.modelMetadata.modelType}
+              telemetryHook={this.props.telemetryHook}
+            />
+          )}
+        {this.state.activeFeatureImportancesOption ===
```
Comment (romanlutz) [inline libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/FeatureImportances.tsx:135]: This isn't really a concern at all, but out of curiosity I'm wondering if it's better/worse or not really any different to do
```
// your solution
{ conditionA && conditionB && <Component1 />}
{ conditionA && !conditionB && <Component2 />}
```
OR
```
{ conditionA && (
    { conditionB ? <Component1 /> : <Component2/> }
)} conditionB && <Component1 />}
{ conditionA && !conditionB && <Component2 />}
```
If you have any thoughts let me know. If not that's fine, just resolve 🙂

File: libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/IndividualFeatureImportanceView/LargeIndividualFeatureImportanceView/LocalImportanceChart.tsx [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+      }
+    ];
+    return {
+      chart: {
+        type: "column"
+      },
+      series: seriesData,
+      title: {
+        text: ""
+      },
```
Comment (romanlutz) [inline libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/IndividualFeatureImportanceView/LargeIndividualFeatureImportanceView/LocalImportanceChart.tsx:248]: Is this intentional? If so, you probably don't need to specify it at all (?)

**Review by romanlutz — COMMENTED** [review]
This is a pretty big PR, so it's kind of hard to review. Will there be tests added at a later point?

File: libs/core-ui/src/lib/util/generateDefaultChartAxes.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  jointDataset: JointDataset,
+  chartType: ChartTypes | OtherChartTypes
+): IGenericChartProps | undefined {
+  if (!jointDataset.hasDataset) {
+    return;
+  }
+  const chartProps: IGenericChartProps = {
+    chartType: chartType ? chartType : ChartTypes.Scatter,
+    colorAxis: {
+      options: {},
```
Comment (tongyu-microsoft) [inline libs/core-ui/src/lib/util/generateDefaultChartAxes.ts:22]: Is `options` needed here?

File: libs/core-ui/src/lib/util/generateDefaultChartAxes.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      property: jointDataset.hasPredictedY
+        ? JointDataset.PredictedYLabel
+        : JointDataset.IndexLabel
+    },
+    xAxis: {
+      options: {},
```
Comment (tongyu-microsoft) [inline libs/core-ui/src/lib/util/generateDefaultChartAxes.ts:28]: same here

File: libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/IndividualFeatureImportanceView/LargeIndividualFeatureImportanceView/LargeIndividualFeatureImportanceView.tsx [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+  };
+
+  private onXSet = (value: ISelectorConfig): void => {
+    if (this.state.chartProps) {
+      const newProps = getNewChartProps(value, true, this.state.chartProps);
+      this.setState({ chartProps: newProps });
+    }
+  };
+
+  private onYSet = (value: ISelectorConfig): void => {
```
Comment (tongyu-microsoft) [inline libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/IndividualFeatureImportanceView/LargeIndividualFeatureImportanceView/LargeIndividualFeatureImportanceView.tsx:234]: the only difference for onXSet and onYSet is `xSet: boolean`, maybe we can combine these two methods into one by passing `xSet: boolean`

File: libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/IndividualFeatureImportanceView/LargeIndividualFeatureImportanceView/LocalImportanceChart.tsx [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+  sortAbsolute: boolean;
+  sortedData: Array<{ [key: string]: number[] | number | undefined }>;
+  topK: number;
+}
+export interface ILocalImportanceData {
+  label: string;
+  value: number;
+}
+
+export const regressionKeyValue = "regressionScore";
```
Comment (tongyu-microsoft) [inline libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/IndividualFeatureImportanceView/LargeIndividualFeatureImportanceView/LocalImportanceChart.tsx]: Do we want to move this to constants.ts?

File: libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/IndividualFeatureImportanceView/LargeIndividualFeatureImportanceView/LocalImportanceChart.tsx [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+  private addScores(): Array<{ [key: string]: number[] }> {
+    const scores: Array<{ [key: string]: number[] }> = new Array(
+      this.props.data?.precomputedExplanations?.localFeatureImportance.scores.length
+    );
+    this.props.data?.precomputedExplanations?.localFeatureImportance.scores.forEach(
+      (score, index) => {
+        scores.push({ [this.props.weightOptions[index + 1]]: score[0] });
+      }
+    );
+    return scores.filter((s) => s);
```
Comment (tongyu-microsoft) [inline libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/IndividualFeatureImportanceView/LargeIndividualFeatureImportanceView/LocalImportanceChart.tsx:207]: what does this filter do? filter undefined s? can we check if score[0] is defined at line 210?

File: libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/IndividualFeatureImportanceView/LargeIndividualFeatureImportanceView/LocalImportanceChart.tsx [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+    _event: React.MouseEvent<HTMLElement, MouseEvent>,
+    checked?: boolean | undefined
+  ): void => {
+    if (checked !== undefined) {
+      this.setState({ sortAbsolute: checked });
+    }
+  };
+
+  private onWeightChange = (option: WeightVectorOption): void => {
+    // add logic to update plot
```
Comment (tongyu-microsoft) [inline libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/IndividualFeatureImportanceView/LargeIndividualFeatureImportanceView/LocalImportanceChart.tsx:265]: is this comment needed?

**Review by tongyu-microsoft — APPROVED** [review]

File: libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/FeatureImportances.tsx (hunk 1 above)
Comment (vinuthakaranth) [reply libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/FeatureImportances.tsx:135]: Thanks Roman, I don't have any thoughts on the second approach. I believe first approach is more readable than the other :D

File: libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/IndividualFeatureImportanceView/LargeIndividualFeatureImportanceView/LocalImportanceChart.tsx (hunk 8 above)
Comment (vinuthakaranth) [reply libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/IndividualFeatureImportanceView/LargeIndividualFeatureImportanceView/LocalImportanceChart.tsx:265]: Great catch! removed it

File: libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/IndividualFeatureImportanceView/LargeIndividualFeatureImportanceView/LocalImportanceChart.tsx (hunk 6 above)
Comment (vinuthakaranth) [reply libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/IndividualFeatureImportanceView/LargeIndividualFeatureImportanceView/LocalImportanceChart.tsx]: moved to constants

File: libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/IndividualFeatureImportanceView/LargeIndividualFeatureImportanceView/LargeIndividualFeatureImportanceView.tsx (hunk 5 above)
Comment (vinuthakaranth) [reply libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/IndividualFeatureImportanceView/LargeIndividualFeatureImportanceView/LargeIndividualFeatureImportanceView.tsx:234]: onXSet and onYset is required by onAccept of AxisConfig. As two AxisConfig is used, we need two different functions to handle. Please let me know if there is a way to merge this. Happy to merge.

File: libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/IndividualFeatureImportanceView/LargeIndividualFeatureImportanceView/LocalImportanceChart.tsx (hunk 2 above)
Comment (vinuthakaranth) [reply libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/IndividualFeatureImportanceView/LargeIndividualFeatureImportanceView/LocalImportanceChart.tsx:242]: Not intentional, I copied a format from other file and this came as is. Removed it.

File: libs/core-ui/src/lib/util/generateDefaultChartAxes.ts (hunk 3 above)
Comment (vinuthakaranth) [reply libs/core-ui/src/lib/util/generateDefaultChartAxes.ts:22]: Yes, in ISelectorConfig interface option is an required param so it is defined as empty object.

File: libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/IndividualFeatureImportanceView/LargeIndividualFeatureImportanceView/LocalImportanceChart.tsx (hunk 7 above)
Comment (vinuthakaranth) [reply libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/IndividualFeatureImportanceView/LargeIndividualFeatureImportanceView/LocalImportanceChart.tsx]: Thanks Tong, I used scores.filter((s) => s) as push was adding items on top of the length of the array and hence was creating empty items. I updated to use [index] and hence filter is not needed now.

**Comment by vinuthakaranth** [comment]
> This is a pretty big PR, so it's kind of hard to review. Will there be tests added at a later point?

I agree, Thank a lot for reviewing!! large data flow is completely separate from existing flow for individual feature importance. All the related files are in separate folder. Yes, tests will be added at later point of time once we have all changes working.

**Pull request merged**